steel/src/verifier.rs

```diff
@@ -0,0 +1,233 @@
+        /// [EvmEnv::into_input]: crate::EvmEnv::into_input
+        pub fn preflight(env: &'a mut HostEvmEnv<D, H, C>) -> Self {
+            Self { env }
+        }
+
+        /// Preflights the commitment verification on the host.
+        pub async fn verify(self, commitment: &Commitment) -> anyhow::Result<()> {
+            log::info!("Executing preflight verifying {:?}", commitment);
+
+            let (id, version) = commitment.decode_id();
+            match version {
+                0 => {
+                    let block_number = validate_block_number(self.env.header().inner(), id)
+                        .context("invalid id")?;
+                    let block_hash = self
+                        .env
+                        .spawn_with_db(move |db| db.block_hash(block_number))
+                        .await?;
+                    ensure!(block_hash == commitment.digest, "invalid digest");
+
+                    Ok(())
+                }
+                1 => {
+                    let beacon_root = self
+                        .env
+                        .spawn_with_db(move |db| BeaconRootsContract::get_from_db(db, id))
+                        .await
+                        .with_context(|| format!("calling BeaconRootsContract({}) failed", id))?;
+                    ensure!(beacon_root == commitment.digest, "invalid digest");
+
+                    Ok(())
+                }
+                v => unimplemented!("Invalid commitment version {}", v),
+            }
+        }
+    }
+
+    impl<D, H, C> HostEvmEnv<D, H, C>
+    where
+        D: Database + Send + 'static,
+    {
+        /// Runs the provided closure that requires mutable access to the database on a thread where
+        /// blocking is acceptable.
+        ///
+        /// It panics if the closure panics.
+        /// This function is necessary because mutable references to the database cannot be passed
+        /// directly to `tokio::task::spawn_blocking`. Instead, the database is temporarily taken out of
+        /// the `HostEvmEnv`, moved into the blocking task, and then restored after the task completes.
+        async fn spawn_with_db<F, R>(&mut self, f: F) -> R
+        where
+            F: FnOnce(&mut ProofDb<D>) -> R + Send + 'static,
+            R: Send + 'static,
+        {
+            // as mutable references are not possible, the DB must be moved in and out of the task
+            let mut db = self.db.take().unwrap();
+
+            let (result, db) = tokio::task::spawn_blocking(|| (f(&mut db), db))
+                .await
+                .expect("DB execution panicked");
+
+            // restore the DB, so that we never return an env without a DB
+            self.db = Some(db);
+
+            result
+        }
+    }
+}
+
+fn validate_block_number(header: &impl EvmBlockHeader, block_number: U256) -> anyhow::Result<u64> {
+    let block_number: u64 = block_number.saturating_to();
+    let diff = header.number().saturating_sub(block_number);
```

Comment:
I feel like this should be `try_into` and `checked_sub`. I suppose these cases would be caught be the subsequent checks (e.g. there is no block hash in the DB for a future block number) but it seems prudent to exist early here IMO.
```suggestion
    let block_number: u64 = block_number.try_into().context("invalid block number >= 2^64")?;
    let diff = header.number().checked_sub(block_number).context("block number is greater than header block number")?;
```

Reply:
Addressed in 18070b4b707a57a93ed2c1529161fbc9defaa4fc
I kept the `saturating_sub` as underflows are caught be the `ensure` in the following line and IMHO produce the better error message.